### PR TITLE
Add support for XDG directories

### LINKED ISSUE #570: Follow XDG conventions on unix-like OS
The XDG conventions are almost universally followed in all CLI tools on Linux & macOS & it would be nice if granted did the same.
At the very least, it should be possible to customize the granted path so it is not `~/.dgranted`.

### PR DESCRIPTION
**Pull request by chris3ware**
### What changed?
Updates have been made to make use of the XDG Base directory specification if the environment variables are set on the users host machine.

config file is stored in `$XDG_CONFIG_HOME`
log file is stored in `$XDG_STATE_HOME`
FrecencyStore is stored in `$XDG_CACHE_HOME`

### Why?
To de-clutter a users `$HOME` directory.
To attempt to resolve #570 

### How did you test it?
Tested locally with `make cli` with the environment variables set and unset. Went through initial configuration, assuming roles and uninstalling on macOS. I can install, setup, assume roles and uninstall with only 1 problem:

With the XDG environment variables set, I see the following error in the log file:

```json
{"level":"debug","ts":"2024-08-07T13:57:19Z","msg":"upserting entry to frecency: open /Users/chris.harrison6/.cache/.dgranted/aws_profiles_frecency: no such file or directory"}
```

I must be missing a step somewhere because `~/.cache/.dgranted` isn't being created. I'm fairly new to go and programming in general so any assistance would be very welcome. No doubt there will be a better way to achieve some of this, but thought I would get the ball rolling :smiley:

### Potential risks
Unknown

### Is patch release candidate?
Unknown

### Link to relevant docs PRs
https://specifications.freedesktop.org/basedir-spec/latest/
https://wiki.archlinux.org/title/XDG_Base_Directory

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -174,8 +176,14 @@ func GrantedConfigFolder() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	// check if the .granted folder already exists
-	return path.Join(home, build.ConfigFolderName), nil
+
+	configDir := filepath.Join(home, build.ConfigFolderName)
+	if xdgConfigDir := os.Getenv("XDG_CONFIG_HOME"); !pathExists(configDir) && xdgConfigDir != "" {
+		// ? Should this be "build.ConfigFolderName" or "granted". Directory names in XDG_CONFIG_HOME usually omit the dot prefix. 
+		configDir = filepath.Join(xdgConfigDir, build.ConfigFolderName)
```
Comment (chrnorm) [inline pkg/config/config.go:183]: Let's use just "granted" here. `build.ConfigFolderName` is currently used internally by our core team to run development builds of Granted using `~/.dgranted` as the config folder.

```suggestion
		configDir = filepath.Join(xdgConfigDir, "granted")
```

File: pkg/config/config.go [hunk 2]
```diff
@@ -187,6 +195,57 @@ func GrantedConfigFilePath() (string, error) {
...
+func GrantedCacheFolder() (string, error) {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	cacheDir := filepath.Join(home, build.ConfigFolderName)
+	if xdgCacheDir := os.Getenv("XDG_CACHE_HOME"); !pathExists(cacheDir) && xdgCacheDir != "" {
+		// ? Should this be "build.ConfigFolderName" or "granted". Directory names in XDG_CACHE_HOME usually omit the dot prefix. 
+		cacheDir = filepath.Join(xdgCacheDir, build.ConfigFolderName)
```
Comment (chrnorm) [inline pkg/config/config.go:207]:
```suggestion
		cacheDir = filepath.Join(xdgCacheDir, "granted")
```

File: pkg/config/config.go [hunk 3]
```diff
@@ -187,6 +195,57 @@ func GrantedConfigFilePath() (string, error) {
...
+func GrantedStateFolder() (string, error) {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	stateDir := filepath.Join(home, build.ConfigFolderName)
+	if xdgStateDir := os.Getenv("XDG_STATE_HOME"); !pathExists(stateDir) && xdgStateDir != "" {
+		// ? Should this be "build.ConfigFolderName" or "granted". Directory names in XDG_STATE_HOME usually omit the dot prefix. 
+		stateDir = filepath.Join(xdgStateDir, build.ConfigFolderName)
```
Comment (chrnorm) [inline pkg/config/config.go:222]:
```suggestion
		stateDir = filepath.Join(xdgStateDir, "granted")
```

**Review by chrnorm — CHANGES_REQUESTED** [review]

File: pkg/config/config.go (hunk 1 above)
Comment (chris3ware) [reply pkg/config/config.go]: New commit pushed with recommended change.

File: pkg/config/config.go (hunk 2 above)
Comment (chris3ware) [reply pkg/config/config.go]: New commit pushed with recommended change.

File: pkg/config/config.go (hunk 3 above)
Comment (chris3ware) [reply pkg/config/config.go]: New commit pushed with recommended change.

**Comment by chris3ware** [comment]
Still seeing the error relating the cache folder. Not sure how to resolve this.

**Review by chrnorm — APPROVED** [review]

**Comment by chrnorm** [comment]
Thankyou @chris3ware 🙌

**Comment by DanCardin** [comment]
Just found granted today, seems great! Any chance of a release soon that'd include this? the latest 0.32 release predates this by a week

**Pull request merged**